Add email template endpoints

joshcanhelp opened the pull request:
- Adding `GET`, `PATCH`, and `POST` for [email templates API](https://auth0.com/docs/api/management/v2#!/Email_Templates/get_email_templates_by_templateName)
- Adding tests for these new methods
- Added a more IDE-friendly method to create an ApiClient
- Added static methods to the ApiTest class to allow for fixture setup

lbalmaceda commented on src/API/Helpers/ApiClient.php:
```diff
@@ -66,4 +67,29 @@ public function __call($name, $arguments) {
+        $builder->withHeaders($this->headers);
+
+        if ( in_array( $method, [ 'patch', 'post', 'put' ] ) ) {
+            $builder->withHeader( new ContentType('application/json') );
```
I know this works for when you need to send a `body` in the request, but probably would be good to check if **all requests** send the `Accept` header so the server knows what you're expecting back in the response.

joshcanhelp replied:
I don't see an `Accept` header used in any of the examples in the docs. I'm wary of just dropping that in if it's not documented.

lbalmaceda commented on src/API/Management/EmailTemplates.php:
```diff
@@ -0,0 +1,77 @@
+     *
+     * @return array
+     *
+     * @throws \Exception - if a 200 response was not returned from the API.
```
Be carefull. Some endpoints return success codes different to `200`, i.e. create connection:

![image](https://user-images.githubusercontent.com/3900123/39211241-6df4c6fa-47e1-11e8-8d92-9f420e37e2a9.png)

While this doesn't apply to email-templates, might be worth to check the rest of the entities if this was a copy-paste of that logic/javadocs.

joshcanhelp replied:
I checked before submitting, all are 200 but good eyes 👍

lbalmaceda commented on tests/API/Management/EmailTemplatesTest.php:
```diff
@@ -0,0 +1,124 @@
+    const EMAIL_TEMPLATE_NAME = 'enrollment_email';
+
+    /**
+     * Management API token with scopes read:email_templates, create:email_templates, update:email_templates
```
while this is true for calling all endpoints, some of them can be called with only 1 of this scopes. So this is not exactly true

joshcanhelp replied:
This is in a test and for the tests here to run, it needs all three.

joshcanhelp commented:
Thanks @lbalmaceda ... this is definitely more documentation than the other methods have but I'm trying to use it as a good template to go back through the others and fill it all out. Your suggestions are 👍

joshcanhelp commented:
@lbalmaceda - Responses to your comment:

- "required scope" - good call, added
- "Some methods require at least some parameters ... validate those cases locally" - I thought about this and was on the fence. You'll always need a template name, sure, but everything else is subject to change, new fields to be added, etc. I can see the utility in checking for required fields now but that's more to update ... and frustrate developers down the road. I added a more clear comment on this but happy to keep talking about how this should be handled. 
- "Template names are limited" - Same general feelings as the above. I started to go down that route and thought that new templates would be a very likely scenario and if we're only whitelisting certain ones then we'd have to stay on top of additions there and add them quickly or someone is going to get a confusing kick-back from the SDK.

lbalmaceda commented:
About local validation: 
- to create a new email-template you require all this properties at least: "template", "body", "from", "subject", "syntax", "enabled", the rest are optional.
- to get/patch an existing email-template you require the "templateName".

About template names, the alternative I proposed (even on my PR) is to pre-define a set of constants that users can use to easily type a template's name. i.e. `const TEMPLATE_ENROLL = "enrollment_email"`

joshcanhelp commented:
> to create a new email-template you require all this properties at least: "template", "body", "from", "subject", "syntax", "enabled", the rest are optional.

For now but one could be added later and now the SDK is out of date. I'm just thinking about that as a pattern for the rest of the SDK. We're duplicating the validation that's happening on the server to save a handful of API calls at the cost of a lot of up-front and maintenance work. 

$0.02

> to get/patch an existing email-template you require the "templateName".

Correct, those are already there. I could require a template name for `create` but then the rest of the fields are required as well.

lbalmaceda commented:
> but one could be added later and now the SDK is out of date

False! If the server adds a new required parameter, the validation that now fails is the remote one and not the local. Yes, SDK is out of date and will need to be updated if you want the validation to detect that locally first. But that's not stopping people from calling this method. TBH I find this really useful! Otherwise every API/Entity would rely on a single generic class that receives a) the method, b) the path and c) the data/body. What's the point on having separate classes for them then? Would like to hear from someone else too (@cocojoe?).


> Correct, those are already there.

But they are not being null-checked. Would it be fine if the request fires with a null template name?

lbalmaceda commented on src/API/Management/EmailTemplates.php:
```diff
@@ -60,18 +107,49 @@ public function patch($templateName, $data)
+        $data = [
+            'template' => (string) $template,
+            'from' => (string) $from,
+            'body' => (string) $body,
```
How does validation work here? And what happens since some of the params have a default value?

joshcanhelp replied:
Anything with a default value is not required by the API, only `template` and `from`. The object you POST needs to have those keys but they can be empty.

`$template` and `$from` are both required for the method (IOW: need to have something passed in) but nothing beyond that. Looking through the rest of the SDK, this is how validation is handled for other endpoints. I could check for `empty()` but it's a bit of an edge case if someone sees that the param is required but passes in an empty string or `null` or something. 

If it was empty, I'd have to `throw` and exception of some kind. Since this is a new method, I could do that even though none of the other ones do and make that a pattern to follow. If that template is empty, the HTTP client would throw an exception anyways so not a complete surprise. 

I'm happy with this as-is but also open to more discussion.

lbalmaceda replied:
I'm looking at the schema definition right now.. You shouldn't be setting values in the SDK if they are missing just because "need to have something passed in". Of course you'll need to pass something, that's why it's required. But let it fail gracefully! If the parameters are required, make the user manually set them, even if they want to set an empty string. The only exception I find for this is the `syntax` field which is there for backwards (read) compatibility and `liquid` should be used always as value when creating new templates. 

Think of receiving an email in your personal inbox which doesn't have a subject set or which doesn't have a body at all. Would that make sense to you as a user of that Application? Don't you want to make sure that users of this SDK use the API well?
